### PR TITLE
Remove 'boost' dependency from Lago

### PR DESCRIPTION
**Pull request by ankur-rc**
Currently, Lago was using algorithms from the Boost Graph Library (BGL), most notably to compute the Minimum Spanning Tree.

This PR attempts to remove the dependency of BGL from Lago by making the following changes:
- Factors out the `kruskal` routine from `SubGraphBuilder` into a free function
- Creates a function `findminimumSpanningTree` in Lago which uses `kruskal`
- Removes all references of BGL within Lago by not including the `graph.h` header

### CODE REVIEW

File: gtsam/base/kruskal-inl.h [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file kruskal-inl.h
+ * @date Dec 31, 2009
+ * @author Frank Dellaert, Yong-Dian Jian
```
Comment (dellaert) [inline gtsam/base/kruskal-inl.h:15]: * @author Frank Dellaert
 * @author Yong-Dian Jian
 * @author Ankur Roy Chowdhury

File: gtsam/base/kruskal.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+namespace gtsam::utils {
+/**
+ * Compute the minimum spanning tree (MST) using Kruskal's algorithm
+ * @param fg Factor graph
+ * @param weights Weights of the edges/factors in the factor graph
+ * @return Edge/factor indices spanning the MST
+ * @note Only binary factors are considered while constructing the spanning tree 
+ * @note The indices of 'weights' should match the indices of the edges in the factor graph 
+ */
+template <class FactorGraph>
```
Comment (dellaert) [inline gtsam/base/kruskal.h:35]: We might want to say

```
template <class FACTOR>
std::vector<size_t> kruskal(const FactorGraph<FACTOR> &fg,
                                  const std::vector<double> &weights);
```

i.e., insist it's a FactorGraph

File: gtsam/base/kruskal.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+std::vector<size_t> kruskal(const FactorGraph &fg,
+                            const std::vector<double> &weights);
+}  // namespace gtsam::utils
+
+#include <gtsam/base/kruskal-inl.h>
```
Comment (dellaert) [inline gtsam/base/kruskal.h:40]: Missing newline at end of file.

File: gtsam/linear/SubgraphBuilder.h [hunk 4]
```diff
@@ -187,4 +185,4 @@ GaussianFactorGraph buildFactorSubgraph(const GaussianFactorGraph &gfg,
 std::pair<GaussianFactorGraph, GaussianFactorGraph> splitFactorGraph(
     const GaussianFactorGraph &factorGraph, const Subgraph &subgraph);
 
-}  // namespace gtsam
+}  // namespace gtsam
```
Comment (dellaert) [inline gtsam/linear/SubgraphBuilder.h:188]: missing newline

File: gtsam/slam/lago.h [hunk 5]
```diff
@@ -62,13 +62,20 @@ GTSAM_EXPORT key2doubleMap computeThetasToRoot(
...
+    /*INPUTS*/const PredecessorMap& tree, const NonlinearFactorGraph& g);
 
 /** Linear factor graph with regularized orientation measurements */
 GTSAM_EXPORT GaussianFactorGraph buildLinearOrientationGraph(
     const std::vector<size_t>& spanningTreeIds,
     const std::vector<size_t>& chordsIds, const NonlinearFactorGraph& g,
-    const key2doubleMap& orientationsToRoot, const PredecessorMap<Key>& tree);
+    const key2doubleMap& orientationsToRoot, const PredecessorMap& tree);
+
+/** Given a "pose2" factor graph, find it's minimum spanning tree.
```
Comment (dellaert) [inline gtsam/slam/lago.h:73]: spelling: its ?

File: tests/testSubgraphSolver.cpp [hunk 6]
```diff
@@ -129,6 +129,7 @@ TEST( SubgraphSolver, constructor3 )
   DOUBLES_EQUAL(0.0, error(Ab, optimized), 1e-5);
 }
 
+
```
Comment (dellaert) [inline tests/testSubgraphSolver.cpp:132]: remove diff

File: gtsam/slam/lago.cpp [hunk 7]
```diff
@@ -211,42 +216,77 @@ static PredecessorMap<Key> findOdometricPath(
...
+    const NonlinearFactorGraph& pose2Graph) {
+  // Compute the minimum spanning tree
+  const auto edgeWeights = std::vector<double>(pose2Graph.size(), 1.0);
+  const auto mstEdgeIndices =
+      utils::kruskal(pose2Graph, edgeWeights);
+
+  // Create a PredecessorMap 'predecessorMap' such that:
+  // predecessorMap[key2] = key1, where key1 is the 'parent' node for key2 in
+  // the spanning tree
+  PredecessorMap predecessorMap;
```
Comment (dellaert) [inline gtsam/slam/lago.cpp:240]: I guess you opted to not add to kruskal?

File: gtsam/linear/SubgraphBuilder.cpp [hunk 8]
```diff
@@ -406,6 +359,8 @@ Subgraph SubgraphBuilder::operator()(const GaussianFactorGraph &gfg) const {
 /****************************************************************/
 SubgraphBuilder::Weights SubgraphBuilder::weights(
     const GaussianFactorGraph &gfg) const {
+  using Weights = std::vector<double>;
```
Comment (dellaert) [inline gtsam/linear/SubgraphBuilder.cpp:362]: hmmm, inside SubgraphBuilder this should be available?

File: gtsam/base/kruskal-inl.h [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+/*****************************************************************************/
+/* sort the container and return permutation index with default comparator */
+inline std::vector<size_t> sortedIndices(const std::vector<double> &src) {
+  const size_t n = src.size();
+  std::vector<std::pair<size_t, double>> tmp;
+  tmp.reserve(n);
+  for (size_t i = 0; i < n; i++) tmp.emplace_back(i, src[i]);
+
+  /* sort */
```
Comment (dellaert) [inline gtsam/base/kruskal-inl.h:39]: // ...

File: gtsam/base/kruskal-inl.h [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+#include <memory>
+#include <vector>
+
+namespace gtsam::utils {
+
+/*****************************************************************************/
+/* sort the container and return permutation index with default comparator */
+inline std::vector<size_t> sortedIndices(const std::vector<double> &src) {
+  const size_t n = src.size();
```
Comment (dellaert) [inline gtsam/base/kruskal-inl.h:34]: What's happenin' here ? (missing comment for this code block :-))

File: gtsam/base/kruskal-inl.h [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+  // Loop over all edges in order of increasing weight.
+  size_t count = 0;
+  for (const size_t index : sortedIndices) {
+    const auto factor = fg[index];
+
+    // Ignore non-binary edges.
+    if (factor->size() != 2) continue;
+
+    auto u = dsf.find(factor->front()), v = dsf.find(factor->back());
```
Comment (dellaert) [inline gtsam/base/kruskal-inl.h:84]: comment

**Review by dellaert — APPROVED** [review]
Very cool ! Will add @lucacarlone as a reviewer so can weigh in with comments, but almost ready to merge after addressing mostly "nits" :-)

File: gtsam/slam/lago.cpp (hunk 7 above)
Comment (ankur-rc) [reply gtsam/slam/lago.cpp:240]: Yes, for the following 2 reasons:
1. I think this function is a bit too specific for Lago and would require a bit more effort to make it more generic.
2. Making `kruskal` return a `PredecessorMap` would mean that we would have to re-extract the edge indices in `SubGraphBuilder` or make `kruskal` return both the `PredecessorMap` and the edge indices.

**Review by dellaert — APPROVED** [review]

**Comment by dellaert** [comment]
will kill all but required CI to unblock my CI experiments in other branch, then will merge

**Comment by dellaert** [comment]
Congrats @ankur-rc on your first gtsam PR :-)

**Pull request merged**